### ISSUE TITLE
Access Denied from IP

### ISSUE THREAD

**Issue by s-pk** [docker-library/ghost#29]
I just used the docs to run an official Ghost image on my computer, but after I get through step two of my blog setup, I get this message at the bottom and cannot move any further 
"Access Denied from url: 10.0.1.13. Please use the url configured in config.js"
That is the current url of my docker machine. Any idea on how to fix this?

**Comment by adam-beck** [comment]
@Skilgarriff are you attempting to access the admin pages using `10.0.1.13:2368`?

**Comment by Skilgarriff** [comment]
I ran the command: docker run --name some-ghost -p 80:2368 -d ghost. So yes, I am attempting to get to the admin pages through 10.0.1.13 but port 80 not 2368. Everything works fine except for 1. If I click the home button it tries to bring me to localhost:2368, and 2. After I register an account, I get a message at the bottom of the page saying access denied from IP 10.0.1.13.

**Comment by squader** [comment]
The issue was reproduced by me on a ubuntu 14.04 machine. After the image got fired up via

`docker run --name some-ghost -p 8080:2368 -d ghost`

an error message states 

> Access Denied from url: (IP). Please use the url configured in config.js.

 when I tried to log in to the backend by calling http://IP:8080/ghost.

**Comment by adam-beck** [comment]
Wow sorry, I'm just now getting notified that you replied. But the way the `config.js` file is set up by default won't allow you to access the admin pages from any URL. You _must_ use `127.0.01` or `localhost`. Is there a reason you are trying to use 10.0.1.13 instead of either of those 2 IP addresses? You can mount your own `config.js` with the proper url configured. In this case, it would probably be under the development section of the config. Changing this would also solve your problems regarding **1**. The links generated by ghost use that url as a reference which is why you are seeing it trying to bring you to `localhost:2368`.

**Comment by AriaFallah** [comment]
@adam-beck In my case, when you run docker on OSX, it doesn't run natively. It gets put into a tiny VM, and runs on a 192.168 IP

**Comment by adam-beck** [comment]
@AriaFallah I may have mispoken when I said you _must_ use `127.0.01` or `localhost`. You can very well copy the `config.example.js` in the container and modify the `development.url` or `development.server.host`/`development.server.port`. I'm not sure which one would actually need changed but I'm assuming just `development.url`. So try changing that to your docker IP.

I'm probably not the best help at this but trying to offer my best guess.

**Comment by adam-beck** [comment]
@AriaFallah what happens when you try accessing `192.168.x.xxx:2368/ghost`?

**Comment by AriaFallah** [comment]
@adam-beck Works just fine. It's the part where it asks me to "get started" by making an account where I get denied.

**Comment by Skilgarriff** [comment]
Would it be possible to make the config.js file (at least the development part of the config) dynamic in that it recognizes the ip of the host it is on and then assigns that to the development.url? Just because the range of IP addresses that can be used for docker-machine is massive and no IP address listed there would really be a one size fits all.

**Comment by yosifkit** [comment]
Yeah, I think the current `config.js` is rather lacking in automatic usefulness.

**Comment by jokesterfr** [comment]
Could we at least have some env to set? Its seems that for now only the NODE_ENV is interpreted (to find the relevant section in config.js).
Setting a `config.js` from a Docker command line is uneasy, however envs can easily be set with `-e`.

We could imagine something like:

 * `-e GHOST_HOST=0.0.0.0`
 * `-e GHOST_PORT=42`

But these shall bypass the defaults set in config.js... If you agree with it, I can implement it.

**Comment by Skilgarriff** [comment]
@jokesterfr Yes, I think that would be very helpful - perhaps also allowing both the production and the development host to be set although they will default to something if not set.

**Comment by jokesterfr** [comment]
Okay I found out a simpler way. No one like to set 10 env vars right?

## TL;DR

```bash
cat <<EOF > config.js
// Ghost Configuration
var path = require('path'),
    config;

config = {
    development: {
        url: 'http://blog.perdu.com',
        database: {
            client: 'sqlite3',
            connection: {
                filename: path.join(process.env.GHOST_CONTENT, '/data/ghost.db')
            },
            debug: false
        },
        server: {
            host: '0.0.0.0',
            port: '2368'
        },
        paths: {
            contentPath: path.join(process.env.GHOST_CONTENT, '/')
        }
    }
};
config.production = config.development;
module.exports = config;
EOF

docker run -p 8080:2368 -v $PWD/config.js:/var/lib/ghost/config.js -d ghost
```

## Naive try

Let's use a custom `config.js`, given across a docker volume. I don't know why I thought volumes could only be folders... (However with *docker compose* you may encounter [some issues](https://github.com/docker/compose/issues/424) with this behaviour)

Here is my `docker run` command:

```bash
docker run -p 8080:2368 -v $PWD/config.js:/var/lib/ghost/config.js -d ghost
```

The container wasn't standing for long, because I experienced issues related to the database:

```bash
Knex:warning - Pool2 - Error: Pool was destroyed
Knex:Error Pool2 - Error: SQLITE_CANTOPEN: unable to open database file
Migrations: Database initialisation required for version 004
Migrations: Creating tables...
Migrations: Creating table: posts

ERROR: Pool is destroyed
```

This is due to the paths I was using in my custom __config.js__ file. I did a naive reuse of my previous (non docker-contained) *config.js* file, and paths just got wrong ! Folks of *docker-library/ghost* just implemented some fallback stuff... Which seems to be poorly documented :p 

## Getting into  docker-library/ghost internals

From what I understood, there are two paths to know about: 

* __$GHOST_SOURCE__: contains all the ghost sources, coming from the official zipball
* __$GHOST_CONTENT__: an empty directory, set as a docker volume to let you customize the `./content` folder used by Ghost. Basically this can contain:
    - __./data__  for the SQLite3 database file
    - __./theme__ to add your custom themes
    - __./images__
    - __./apps__
    - .... and also __index.js__ as an exception

And if you read carefully the `docker-entrypoint.sh` magic, you get that the fallback works as follow:

1. Try to symlink `$GHOST_CONTENT/config.js` to  `$GHOST_SOURCE/config.js`
2. If no custom config is found, get the sample from the official GHOST platform (config.example.js), replacing the `127.0.0.1` with `0.0.0.0` and `path/to/content` by `process.env.GHOST_CONTENT`

Replacing the `127.0.0.1` ip address is a good idea, Node.js HTTP server is listening on an IP that's not used by [docker networking](https://docs.docker.com/v1.8/articles/networking). Instead `0.0.0.0` will accept any IPv4 address (see [Node.js documentation](https://nodejs.org/api/http.html#http_server_listen_port_hostname_backlog_callback)).

The replacement of `path/to/content` by `process.env.GHOST_CONTENT` is the tricky bit added by *docker-library/ghost*, because there is no `/var/lib/ghost/content` inside the container, and the `config.js` file is supposed to be at the same level as the `./data` or `./theme` folders. 

How do you guys think this can be simplified?

**Comment by jdc0589** [comment]
@adam-beck same issue here. This needs to be simplified via an environment variable or something.

**Comment by Skilgarriff** [comment]
@adam-beck @jdc0589 Like [this?](https://github.com/ZZROTDesign/alpine-ghost) Basically this exact image, but with environment variables built into the original config.js (you can still manually override this in the same manner as this image)

**Comment by jdc0589** [comment]
@Skilgarriff wow...that was easy. yes, exactly like that.
Thanks.

**Comment by adam-beck** [comment]
@Skilgarriff I really like that approach. Curious as to why you aren't using `0.0.0.0` for DEV_DOMAIN in the `docker-compose.example.yml` though?

**Comment by Skilgarriff** [comment]
@adam-beck You want DEV_DOMAIN to be the IP of your docker-machine. So for me, my docker-machine ip is something like 198.280.xx.xx:2368. I put that in for DEV_DOMAIN Which tells ghost how to navigate/verify the domain when trying to signup for your admin page.

**Comment by tianon** [comment]
With Ghost 1.x, everything can be configured via environment variables now. :+1:

See https://docs.ghost.org/v1.0.0/docs/config#section-running-ghost-with-config-env-variables for more details.

Additionally, the default 1.x configuration is now a production config instead of development.
